lightning/src/offers/invoice.rs

```diff
@@ -2235,15 +2236,25 @@ mod tests {
 				.build_and_sign()
 				.unwrap();
 
-		if let Err(e) = invoice_request
+		let verified_request = invoice_request
 			.clone()
 			.verify_using_recipient_data(nonce, &expanded_key, &secp_ctx)
-			.unwrap()
-			.respond_using_derived_keys_no_std(payment_paths(), payment_hash(), now())
-			.unwrap()
-			.build_and_sign(&secp_ctx)
-		{
-			panic!("error building invoice: {:?}", e);
+			.unwrap();
+
+		match verified_request {
+			InvoiceRequestVerifiedFromOffer::DerivedKeys(req) => {
+				let invoice = req
+					.respond_using_derived_keys_no_std(payment_paths(), payment_hash(), now())
+					.unwrap()
+					.build_and_sign(&secp_ctx);
+
+				if let Err(e) = invoice {
+					panic!("error building invoice: {:?}", e);
+				}
+			},
+			InvoiceRequestVerifiedFromOffer::ExplicitKeys(_) => {
+				panic!("expected invoice request with keys");
```

Comment:
This match arm with a panic comes back a few times. It does feel a bit anti-pattern'ish. If we go for strong types to eliminate runtime checks, I wouldn't expect other runtime checks being added.